src/dynamic_honey_badger/dynamic_honey_badger.rs

```diff
@@ -34,8 +36,120 @@ pub struct DynamicHoneyBadger<C, N: Rand> {
     pub(super) honey_badger: HoneyBadger<InternalContrib<C, N>, N>,
     /// The current key generation process, and the change it applies to.
     pub(super) key_gen_state: Option<KeyGenState<N>>,
-    /// A queue for messages from future epochs that cannot be handled yet.
-    pub(super) incoming_queue: Vec<(N, Message<N>)>,
+    /// `HoneyBadger` messages that couldn't be handled yet by remote nodes. Each key to the map
+    /// contains the target node ID, the Dynamic Honey Badger start epoch and the embedded Honey
+    /// Badger epoch of a vector of messages stored as the value of that key.
+    pub(super) outgoing_queue_hb: BTreeMap<(N, DynamicEpoch), Vec<Message<N>>>,
+    /// Messages created by `DynamicHoneyBadger` that cannot be handled yet by remove nodes. The
+    /// keys are pairs of the node ID and the start epoch.
+    pub(super) outgoing_queue_dhb: BTreeMap<(N, u64), Vec<Message<N>>>,
+    /// Known current epochs of remote nodes.
+    pub(super) remote_epochs: BTreeMap<N, DynamicEpoch>,
+    /// Nodes which are being added using a `Change::Add` command. The command is ongoing and hasn't
+    /// been finished yet.
+    pub(super) nodes_being_added: BTreeSet<N>,
+}
+
+pub type Step<C, N> = messaging::Step<DynamicHoneyBadger<C, N>>;
+
+impl<C, N> Step<C, N>
+where
+    C: Contribution + Serialize + for<'r> Deserialize<'r>,
+    N: NodeIdT + Serialize + for<'r> Deserialize<'r> + Rand,
+{
+    /// Removes and returns any messages that are not yet accepted by remote nodes according to the
+    /// mapping `remote_epochs`. This way the returned messages are postponed until later, and the
+    /// remaining messages can be sent to remote nodes without delay.
+    fn defer_messages<'i, I>(
```

Comment:
I've moved the method to `messaging.rs` temporarily and added a few function arguments to make it generic enough to be used be `HoneyBadger` and `DynamicHoneyBadger`. It can be moved to a separate module. I'll pull the queue structures out of the algorithms.